pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaManaged.hpp

```diff
@@ -1671,5 +2524,105 @@ class StateVectorCudaManaged
+                                     const std::vector<bool> &ctrls_values,
+                                     bool use_adjoint = false) {
+        void *extraWorkspace = nullptr;
+        std::size_t extraWorkspaceSizeInBytes = 0;
+        int nIndexBits = BaseType::getNumQubits();
+
+        cudaDataType_t data_type;
+        custatevecComputeType_t compute_type;
+
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type = CUDA_C_64F;
+            compute_type = CUSTATEVEC_COMPUTE_64F;
+        } else {
+            data_type = CUDA_C_32F;
+            compute_type = CUSTATEVEC_COMPUTE_32F;
+        }
+
+        auto ctrlsInt = NormalizeCastIndices<std::size_t, int>(
+            ctrls, BaseType::getNumQubits());
+        auto tgtsInt = NormalizeCastIndices<std::size_t, int>(
+            tgts, BaseType::getNumQubits());
+        auto ctrls_valuesInt =
+            Pennylane::Util::cast_vector<bool, int>(ctrls_values);
+
+        std::reverse(tgtsInt.begin(), tgtsInt.end());
+        std::reverse(ctrlsInt.begin(), ctrlsInt.end());
+        std::reverse(ctrls_valuesInt.begin(), ctrls_valuesInt.end());
+
+        custatevecIndex_t *permutation_data =
+            permutation.empty() ? nullptr : permutation.data();
+
+        // check the size of external workspace
+        PL_CUSTATEVEC_IS_SUCCESS(
+            custatevecApplyGeneralizedPermutationMatrixGetWorkspaceSize(
+                /* custatevecHandle_t */ handle_.get(),
+                /* cudaDataType_t */ data_type,
+                /* const uint32_t */ nIndexBits,
+                /* custatevecIndex_t*  */ permutation_data,
+                /* const void* */ diagonals,
+                /* cudaDataType_t */ data_type,
+                /* const int32_t* */ tgtsInt.data(),
+                /* const uint32_t */ tgtsInt.size(),
+                /* const uint32_t */ ctrlsInt.size(),
+                /* std::size_t* */ &extraWorkspaceSizeInBytes));
+
+        PL_CUDA_IS_SUCCESS(cudaStreamSynchronize(
+            BaseType::getDataBuffer().getDevTag().getStreamID()));
+
+        // allocate external workspace if necessary
+        // LCOV_EXCL_START
+        if (extraWorkspaceSizeInBytes > 0) {
+            PL_CUDA_IS_SUCCESS(
+                cudaMalloc(&extraWorkspace, extraWorkspaceSizeInBytes));
+        }
+        // LCOV_EXCL_STOP
+
+        // apply gate
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecApplyGeneralizedPermutationMatrix(
+            /* custatevecHandle_t */ handle_.get(),
+            /* void* */ BaseType::getData(),
+            /* cudaDataType_t */ data_type,
+            /* const uint32_t */ nIndexBits,
+            /* custatevecIndex_t*  */ permutation_data,
+            /* const void* */ diagonals,
+            /* cudaDataType_t */ data_type,
+            /* const int32_t */ use_adjoint,
+            /* const int32_t* */ tgtsInt.data(),
```

Comment:
Is it possible to make use of the target and control arguments here to avoid creating diagonal elements for the controls? If it works, it will boost the performance at large for certain cases.

Reply:
Another thought: take Controlled-RX as an example, can we use `custatevecApplyPauliRotation` and set the angle as `Pi/2` to represent the generator?

Reply:
So say we want the generator of controlled-RX (|0><0|xI + |1><1|xRX), which is |1><1|xPauliX, which will be
```
0 0 0 0 
0 0 0 0
0 0 0 1
0 0 1 0
```
I don't think you can get this matrix with PauliRotation? The main issue is the block of zero in the complement of the control (so the |0><0| i.e. top left corner block)